packages/workbox-background-sync/src/lib/QueueDb.ts

```diff
@@ -0,0 +1,149 @@
+/*
+  Copyright 2018 Google LLC
+
+  Use of this source code is governed by an MIT-style
+  license that can be found in the LICENSE file or at
+  https://opensource.org/licenses/MIT.
+*/
+
+import {openDB, DBSchema, IDBPDatabase} from 'idb';
+import {RequestData} from './StorableRequest.js';
+
+interface QueueDBSchema extends DBSchema {
+  requests: {
+    key: number;
+    value: UnidentifiedQueueStoreEntry;
+    indexes: {queueName: string};
+  };
+}
+
+const DB_VERSION = 3;
+const DB_NAME = 'workbox-background-sync';
+const OBJECT_STORE_NAME = 'requests';
+const INDEXED_PROP = 'queueName';
+
+export interface UnidentifiedQueueStoreEntry {
+  requestData: RequestData;
+  timestamp: number;
+  id?: number;
+  queueName?: string;
+  metadata?: object;
+}
+
+export interface QueueStoreEntry extends UnidentifiedQueueStoreEntry {
+  id: number;
+}
+
+/**
+ * A class to interact directly with IndexedDB to store a retrieve queue requests.
+ *
+ * @private
+ */
+
+export class QueueDb {
+  private _db: IDBPDatabase<QueueDBSchema> | null = null;
+
+  /**
+   * Opens a connection to IDBDatabase.
+   *
+   * @private
+   */
+  private async open() {
```

Comment:
Usage might be cleaner if you have this method always return `this._db`, and name it `getDB()` or something like that.

That way you wouldn't have to put the `!` after `this._db!` all over the place.

Reply:
Agreed, fixed it.